### DISCUSSION TITLE
Naming row fields without explicit type specification

### DISCUSSION THREAD

**Discussion by EvgSkv** [trinodb/trino#7758] [Q&A] [answered]
Is there a way to name fields of a row without specifying their types?
Like this
```sql
select cast(row(1,'hello') as row(x bigint, y varchar));
```
but if I don't want to specify `bigint` and `varchar`?

It seems that it could be technically possible, as Trino can operate on unnamed rows inferring type from the context.
Is there a way to do so?

I'm asking because we are generating SQL and do not have type information at the moment.

**Comment by EvgSkv** [comment]
An answer that this is not possible would also help :-)

**Reply by martint** [reply]
This is currently not possible. I'm looking at the SQL specification to see if there's anything we're missing that would allow this, but there doesn't seem to be any standard functionality for this that we could implement.

**Reply by martint** [reply]
One option, which unfortunately is not currently supported in Trino, would be to use a scalar subquery that produces a ROW type:

```sql
SELECT (SELECT 1 AS x, 2 AS y)
```

**Reply by martint** [reply]
I've filed https://github.com/trinodb/trino/issues/7773

**Reply by martint** [reply, marked as answer]
And here's the implementation: https://github.com/trinodb/trino/pull/7774

```
trino> SELECT typeof((SELECT 1 AS x, 'hello' AS y)), (SELECT 1 AS x, 'hello' AS y);
            _col0             |     _col1
------------------------------+----------------
 row(x integer, y varchar(5)) | {x=1, y=hello}
(1 row)
```

**Reply by EvgSkv** [reply]
Wow. Thank you very much for doing this! This hugely helps my use-case.

**Reply by johnedquinn** [reply]
Interesting, is there any way to write a subquery consisting of a single projection item to still be of a ROW type? AKA to not be coerced into a scalar?
```sql
SELECT (SELECT 1 AS x)
```

I'd also like to create a ROW with named fields without having to specify its type.
